### ISSUE TITLE
semantic partial eq between graphql documents

### ISSUE THREAD

**Issue by lrlna** [apollographql/apollo-rs#356]
There have been a few asks to be able to check whether two documents are semantically the same. Here are some things that I think would not matter when checking for equality:
- order of definitions in a document
- order of fields, enum values, union members
- order of argument definitions, variables, input values
- order of directives

Things like a definition having an extra field, or directive, or argument will result in inequality. 

We'd probably want to implement this first on the parser, but then definitely extend to the compiler's HIR. Even implementing a super small formatting example with the compiler needed some equality checks. I think this would require a partial eq implementation on pretty much every one of our AST nodes, and I don't think we can generate them with the existing `xtask` codegen task and have to manage it manually.

**Comment by yanns** [comment]
Yes, that would be great.
I'm trying to compare 2 documents for test purpose, and if I compare the `.to_string()`, I need to handle some cases like:
```
<type Review @join__type(graph: REVIEW) {
>type Review
>   @join__type(graph: REVIEW) {
```

**Comment by SimonSapin** [comment]
> We'd probably want to implement this first on the parser, but then definitely extend to the compiler's HIR.

If things like ordering are not significant, this comparison makes a lot more sense to me to have on the HIR than on the AST since the HIR already uses maps instead of vectors in some places (and probably should in more places)

**Comment by Geal** [comment]
Why wouldn't order be significant? For queries (they're a type of document for apollo-compiler, right?), field order has a meaning

**Comment by lrlna** [comment]
ordering is not significant for 99% of things, other than the ordering of Directives on their locations. that is, the type here is semantically different

```graphql
type Query {
  name: String @a @b 
}
```
from this one
```graphql
type Query {
  name: String @b @a
}
```

**Comment by lrlna** [comment]
> field order has a meaning

ohhh yea that's another one!

if you move definitions around, ordering doesn't matter

**Comment by SimonSapin** [comment]
Order of selections in a selection set (such as fields in a query) is meaningful. But field order in an object type definition is not, is it?

**Comment by Geal** [comment]
Right, I was talking about selections, sorry for messing up the vocabulary
